A new tracker for Weights & Biases

migalkin opened the pull request:
This PR adds a new ResultTracker for [Weights & Biases](http://wandb.ai).

Similarly to the MLFlowResultTracker, this one allows to track experimental results and metrics via WANDB (which coolness we could discuss in another PR 😆 ).  

The only requirement for a user is to have a wandb account and perform `wandb login <api_key>` before running pykeen, so wandb would recognize a user and a selected project to write logs to. 

Tested with @mali-git on a local pykeen installation

migalkin committed:
A new tracker for Weights & Biases

mali-git committed:
Remove functions commented out, and pass flake8

mali-git committed:
Add missing punctuation

mali-git committed:
Adapt error message

cthoyt commented on src/pykeen/trackers.py:
```diff
@@ -85,3 +86,44 @@ def log_params(self, params: Dict[str, Any], prefix: Optional[str] = None) -> No
+        """
+        import wandb as _wandb
+        self.wandb = _wandb
+        assert project_name is not None, "Weights & Biases requires a project name."
```
throw a proper exception, asserts are usually discarded at runtime

mali-git replied:
Done in edf628828faf35eb2763980b8e08910c82934d3f.

cthoyt commented on src/pykeen/trackers.py:
```diff
@@ -85,3 +86,44 @@ def log_params(self, params: Dict[str, Any], prefix: Optional[str] = None) -> No
+        prefix: Optional[str] = None,
+    ) -> None:  # noqa: D102
+        metrics = flatten_dictionary(dictionary=metrics, prefix=prefix)
+        self.wandb.log(metrics)
```
excellent and elegant!

cthoyt commented on src/pykeen/trackers.py:
```diff
@@ -85,3 +86,44 @@ def log_params(self, params: Dict[str, Any], prefix: Optional[str] = None) -> No
+    def log_params(self, params: Dict[str, Any], prefix: Optional[str] = None) -> None:  # noqa: D102
+        params = flatten_dictionary(dictionary=params, prefix=prefix)
+        for k, v in params.items():
+            self.wandb.config[k] = v
```
this is a dict? you can just do

```python
self.wand.config.update(params)
```

migalkin replied:
Good call, done in 6ff2ef2fbc7b7a9ba51dc003d39d32589237dc7f

cthoyt commented on src/pykeen/trackers.py:
```diff
@@ -85,3 +86,44 @@ def log_params(self, params: Dict[str, Any], prefix: Optional[str] = None) -> No
+        if experiment_name is not None:
+            self.wandb.init(project=self.project_name, name=experiment_name)
+        else:
+            self.wandb.init(project=self.project_name)
```
what's the default value for `name` in the `init()` function?

mali-git replied:
Adressed in 66e1f95fd031e7135c528047955c907341edcf1b.

cthoyt commented on src/pykeen/trackers.py:
```diff
@@ -85,3 +86,44 @@ def log_params(self, params: Dict[str, Any], prefix: Optional[str] = None) -> No
+    """A tracker for Weights & Biases. Note that you have to perform wandb login beforehand."""
+
+    def __init__(
+            self,
```
please fix indentation (only 4 spaces more than the def block)

mali-git replied:
Done in 0894bc36d0621671250f0f6b688e49e4e1bea6ed.

cthoyt committed:
Add stub for wandb

[skip ci]

cthoyt commented on docs/source/tutorial/using_wandb.rst:
```diff
@@ -0,0 +1,3 @@
+Using Weights and Biases
+========================
+`Weights and Biases <http://wandb.ai/>`_ is a...
```
@migalkin @mali-git here's where the tutorial should go

migalkin replied:
Done in 37f61546fbeb5ac150dc923b69e733a2ed1b4574

cthoyt committed:
Update pipeline.py

[skip ci]

cthoyt commented on src/pykeen/pipeline.py:
```diff
@@ -659,6 +659,9 @@ def pipeline(  # noqa: C901
             experiment_name=mlflow_experiment_name,
         )
+
+    # TODO add additional logic for wandb tracking
```
Add additional keyword arguments to the pipeline as well as logic here to instantiate the WANDBResultTracker

cthoyt replied:
Let's merge #69 first and then I will take care of this for you

cthoyt replied:
Okay this is taken care of

cthoyt committed:
Merge branch 'master' into wandb_tracker

cthoyt committed:
Fix docstring

cthoyt committed:
Update README.md

cthoyt committed:
Add wandb to readme

mali-git committed:
Replace assertion with specific check.

mali-git committed:
Fix indention

mali-git committed:
Simplify code

migalkin committed:
A tutorial page on using WANDB Tracker

cthoyt committed:
Merge branch 'master' into wandb_tracker

migalkin committed:
A more elegant wandb config update

cthoyt committed:
Fix typo in mlflow tutorial

cthoyt committed:
Clean up code

- fix spacing in docstring
- use more specific exceptions
- use more succinct variable names

cthoyt committed:
Update tutorial

- Conform to doc8 rules
- make more structured

cthoyt committed:
Automatically assign random seed

cthoyt committed:
Update .gitignore

cthoyt commented on docs/source/tutorial/using_wandb.rst:
```diff
@@ -0,0 +1,64 @@
+        result_tracker='wandb',
+        result_tracker_kwargs=dict(
+            project='pykeen_project',
+            experiment='new run',
```
All pipelines will be passed the same experiment. What happens? Do they get overwritten?

migalkin replied:
No, there will be N runs with the same name but different hashes (unique ids). But generally `experiment` is advisable to use for single runs, not for batches of multiple runs.

cthoyt committed:
Update pipeline.py

cthoyt commented on src/pykeen/pipeline.py:
```diff
@@ -666,7 +666,7 @@ def pipeline(  # noqa: C901
         model_kwargs = {}
     model_kwargs.update(preferred_device=device)
-    model_kwargs.setdefault('random_seed', NoRandomSeedNecessary)
+    model_kwargs.setdefault('random_seed', random_seed)
```
@mberr @mali-git I was definitely the one who implemented this part of the pipeline but for the life of me I can't remember if there was a reason for NoRandomSeedNecessary other than making the HPO not emit a warning message... if not then let me know if you think there are reprecussions for setting this in the config *post facto*

mali-git replied:
If I remember correctly, the reason was that we already set the random seed in:
https://github.com/pykeen/pykeen/blob/22f8815fe6520dc89f786e43d5cb3682abe1113d/src/pykeen/pipeline.py#L642

If we pass `random_seed` to the model, we set it again in:
https://github.com/pykeen/pykeen/blob/22f8815fe6520dc89f786e43d5cb3682abe1113d/src/pykeen/models/base.py#L247

cthoyt replied:
correct, but that only happens if there is no random seed already set so there's both the possibility to do this through the pipeline and so you could roll your own pipeline

cthoyt reviewed (approved):
Besides finalizing the two minor comments tomorrow, this is ready to go! Thanks @migalkin

cthoyt committed:
Merge branch 'master' into wandb_tracker

cthoyt committed:
Update tutorial
